frontends/p4/moveDeclarations.cpp

```diff
@@ -109,24 +109,41 @@ const IR::Node *MoveInitializers::preorder(IR::P4Parser *parser) {
             }
         }
     }
+
     if (someInitializers) {
+        // Check whether any of the parser's states loop back to the start state.
```

Comment:
Isn't there a class which builds the reachability graph for parser states?

Reply:
I'm not sure, but that might help. @fruffy @vlstill ?

Reply:
./frontends/p4/parserCallGraph.h

Reply:
`CallGraph::reachable(start, reachable)` will always include `start` in the set of `reachable` states, so this will not really help simplify things.

Reply:
you can check the set of in-edge of the start state

Reply:
@mihaibudiu My initial attempt to use `CallGraph::isCallee()` failed: https://github.com/p4lang/p4c/commit/4b0ec77109883272b77296e6b742ed01b5051c6a
Please let me know if I'm doing something wrong

Reply:
how did it fail?

Reply:
shouldn't the tests fail if the output is incorrect?

Reply:
No, because I updated them so that you could easily see the changes in the diff. They are passing, but the differences don't look correct.

Reply:
if you are confident that the previous tests were correct, you probably should not have updated the outputs.
this makes it much harder to figure out what the problems are since you have to compare files between commits.

Reply:
but frankly it would be useful to understand why the results are wrong. If there is a bug in the callgraph, it would be nice to know that. If the meaning of the isCalle() function is different than expected, it should be documented. If the rule for splitting is wrong, it would be good to know that too.

Reply:
@mihaibudiu Personally, I find it easier to initially analyze changes by looking at git diffs, which is why I shared the link to the diff with the C++ changes as well as corresponding updated tests outputs. If you prefer to initially analyze changes by checking out the changes locally, running the tests, and inspecting the tests failures locally, you can easily do this by recreating the changes minus tests outputs updates by fetching my fork and using `git checkout` commands.

Reply:
I am no longer working on this project, but I will try to find time to fix bugs that I am responsible for.

Reply:
@mihaibudiu For example: https://github.com/p4lang/p4c/commit/4b0ec77109883272b77296e6b742ed01b5051c6a#diff-cd2b40893d904a1015f5d71d39014e909e148f5db324e0ebb77484e99c3cc6c7R2-R7

Split is expected in this case because `start` is reachable from other states. So either I'm using `ParserCallGraph` incorrectly, or it's not working properly.

> I am no longer working on this project, but I will try to find time to fix bugs that I am responsible for.

I did not realize you were no longer working on this project. Please do not feel obligated to fix even the bugs you are responsible for. But any help is of course appreciated :)